src/tiledb/cloud/utilities/profiler.py

```diff
@@ -0,0 +1,194 @@
+import subprocess
+import threading
+import time
+from typing import Optional
+
+import numpy as np
+
+import tiledb
+from tiledb.cloud.utilities import max_memory_usage
+
+
+def create_log_array(uri: str):
+    """
+    Create an array to hold log events.
+
+    :param uri: array URI
+    """
+
+    int_fl = tiledb.FilterList(
+        [
+            tiledb.DoubleDeltaFilter(),
+            tiledb.BitWidthReductionFilter(),
+            tiledb.ZstdFilter(),
+        ]
+    )
+    ascii_fl = tiledb.FilterList([tiledb.ZstdFilter()])
+
+    max_time = np.iinfo(np.uint64).max - 1
+    d0 = tiledb.Dim(
+        name="time_ms", dtype=np.uint64, domain=(0, max_time), filters=int_fl
+    )
+    dom = tiledb.Domain([d0])
+
+    a0 = tiledb.Attr(name="id", dtype="ascii", filters=ascii_fl)
+    a1 = tiledb.Attr(name="op", dtype="ascii", filters=ascii_fl)
+    a2 = tiledb.Attr(name="data", dtype="ascii", filters=ascii_fl)
+    a3 = tiledb.Attr(name="extra", dtype="ascii", filters=ascii_fl)
+
+    schema = tiledb.ArraySchema(
+        domain=dom,
+        sparse=True,
+        attrs=[a0, a1, a2, a3],
+        offsets_filters=int_fl,
+        allows_duplicates=True,
+    )
+
+    schema.check()
+    tiledb.Array.create(uri, schema)
+
+
+def write_log_event(
+    uri: str,
+    id: str,
+    op: Optional[str] = "",
+    data: Optional[str] = "",
+    extra: Optional[str] = "",
+):
+    """
+    Write an event to the log array.
+
+    :param uri: array URI
+    :param id: event id
+    :param op: event operation, defaults to ""
+    :param data: event data, defaults to ""
+    :param extra: event extra data, defaults to ""
+    """
+
+    t_now_ms = time.time() * 1000
+    with tiledb.open(uri, "w") as A:
+        A[t_now_ms] = {
+            "id": [id],
+            "op": [op],
+            "data": [data],
+            "extra": [extra],
+        }
+
+
+class Profiler(object):
+    """
+    A simple profiler to log start/finish events with CPU and memory usage.
+    """
+
+    def __init__(
+        self,
+        uri: str,
+        id: str,
+        trace: bool = False,
+    ):
+        """
+        Create a profiler object.
+
+        :param uri: event log array URI
+        :param id: profiler id, written to event id
+        :param trace: enable trace logging, defaults to False
+        """
+
+        self.uri = uri
+        self.id = id
+        self.trace = trace
+        self.array = tiledb.open(uri, "w")
+
+        # Log node start time and system info
+        node_id = open("/proc/sys/kernel/random/boot_id").read().strip()
+        cmd = "uptime && free -h && uname -a && lscpu"
```

Comment:
Would it be better to run these individually so that (a) we're not using `shell=True` and (b) we get the data back in a more structured manner?